### PR TITLE
Close IO streams after the plugin has finished reading and writing

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Closes IO streams after the plugin has finished reading and writing, in a `finally` block, which is guaranteed to be executed last.

## Changeset

- Altered existing use of `Closeable` values so that `close` is called in a try-finally block. 
- Added catch block where this was previously missing, that logs a warning

## Tests

Ran mazerunner scenarios.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/main/groovy/com/bugsnag/android/gradle/BugsnagManifestTask.groovy [hunk 1]
```diff
@@ -51,10 +51,20 @@ class BugsnagManifestTask extends BugsnagVariantOutputTask {
...
 
             // Write the manifest file
-            def writer = new FileWriter(manifestPath)
-            def printer = new XmlNodePrinter(new PrintWriter(writer))
-            printer.preserveWhitespace = true
-            printer.print(xml)
+            FileWriter writer = null
+
+            try {
+                writer = new FileWriter(manifestPath)
```
Comment (artem-zinnatullin) [inline src/main/groovy/com/bugsnag/android/gradle/BugsnagManifestTask.groovy]: can't you use try-with-resources here?

Comment (fractalwrench) [reply src/main/groovy/com/bugsnag/android/gradle/BugsnagManifestTask.groovy]: This repo currently targets Java 1.6 at the moment, so unfortunately not without some changes.

I believe the reasoning behind targeting 1.6 is that we compile the local Gradle's distribution of Groovy, which as of [Gradle 4.0](https://docs.gradle.org/4.0/release-notes.html#groovy-upgraded-to-2.4.11) is Groovy 2.4.X. This version of Groovy still supports [1.6 JVMs](http://www.groovy-lang.org/download.html#requirements).

I'll raise an internal ticket to discuss bumping our compatibility to 1.7, as I'd hope most users of Android Studio are using a modern JDK.

Comment (artem-zinnatullin) [reply src/main/groovy/com/bugsnag/android/gradle/BugsnagManifestTask.groovy]: ah I see 👍

File: src/main/groovy/com/bugsnag/android/gradle/BugsnagProguardConfigTask.groovy [hunk 2]
```diff
@@ -41,10 +41,17 @@ class BugsnagProguardConfigTask extends DefaultTask {
...
+        FileWriter fr = null
+
+        try {
+            fr = new FileWriter(file.path)
+            fr.write(PROGUARD_CONFIG_SETTINGS)
+            fr.write("\n")
+        } catch (Exception e) {
+            project.logger.warn("Failed to write Bugsnag ProGuard settings", e)
+        } finally {
+            fr.close()
```
Comment (kattrali) [inline src/main/groovy/com/bugsnag/android/gradle/BugsnagProguardConfigTask.groovy]: Shouldn't there be a null check here, in case the writer was not initialized?

**Comment by Pezzah** [comment]
Should we also close this reader https://github.com/bugsnag/bugsnag-android-gradle-plugin/pull/126/files#diff-ff274d12fbfb73a5d6888d7dc318c167R72

**Comment by fractalwrench** [comment]
Thanks, good catch! I've updated the PR to close the additional reader.
